fix(transcode): conditionally fix VAAPI filter order for HEVC on Haswell

andresichelero opened the pull request:
Places 'format=nv12' before 'hwupload' in the VAAPI video filter chain only when the input codec is HEVC and hardware decoding is disabled.

This resolves filtergraph initialization errors (`Impossible to convert between the formats supported by the filter 'Parsed_hwupload_0' and the filter 'auto_scale_0'`) and i965 driver assertion failures (`Assertion 'obj_surface->fourcc == VA_FOURCC_NV12' failed`) observed on Intel Haswell GPUs (e.g., HD 4400) when transcoding HEVC input streams (both 8-bit and potentially 10-bit) to H.264 output using VAAPI encoding via the software decoding path.

By ensuring the pixel format is set to NV12 *before* the frame is uploaded to the GPU surface via `hwupload`, the subsequent VAAPI filters (like `scale_vaapi`) and the `h264_vaapi` encoder receive the hardware surface in the expected format, preventing the previously observed crashes.

The default filter logic and order are retained for other input codecs (e.g., H.264) and when hardware decoding is active, minimizing potential regressions.

Addresses issues discussed in GitHub issue #18103

Commit:
fix(transcode): conditionally fix VAAPI filter order for HEVC (tested on Haswel) l
Places 'format=nv12' before 'hwupload' in the VAAPI video filter chain only when the input codec is HEVC and hardware decoding is disabled. This resolves filtergraph/driver issues observed on Intel Haswell GPUs when transcoding HEVC->H.264 VAAPI via software decode, ensuring the encoder receives the expected NV12 surface. Retains default filter logic for other codecs. Addresses issues discussed in GitHub issue #18103

mertalev commented on server/src/utils/media.ts:
```diff
@@ -781,7 +781,12 @@ export class VaapiSwDecodeConfig extends BaseHWConfig {
   getFilterOptions(videoStream: VideoStreamInfo) {
     const options = this.getToneMapping(videoStream);
-    options.push('hwupload=extra_hw_frames=64');
+    if (videoStream.codecName === VideoCodec.HEVC) {
+      options.push('format=nv12');
+      options.push('hwupload=extra_hw_frames=64');
+    } else {
+      options.push('hwupload=extra_hw_frames=64');
+    }
     if (this.shouldScale(videoStream)) {
       options.push(`scale_vaapi=${this.getScaling(videoStream)}:mode=hq:out_range=pc:format=nv12`);
     }
```
Whether to set the pixel format isn't really related to whether the input is HEVC. Also,`scale_vaapi` actually takes care of the format too, so I think it's better to do something like this (not tested):
```suggestion
    if (this.shouldScale(videoStream)) {
      options.push('hwupload=extra_hw_frames=64', `scale_vaapi=${this.getScaling(videoStream)}:mode=hq:out_range=pc:format=nv12`);
    } else {
      options.push('format=nv12', 'hwupload=extra_hw_frames=64');
    }
```
The same is true of QSV, so it should receive the same change.

mertalev reviewed (commented):
Would you be able to add unit tests for QSV and VAAPI in `media.service.spec.ts` to cover this case? Specifically, there needs to be a filter to convert to nv12 when the scale filter is absent.

andresichelero committed:
Update server/src/utils/media.ts

- HEVC is not related to format, removed condition
- scale_vaapi takes care of the format

Co-authored-by: Mert <101130780+mertalev@users.noreply.github.com>

andresichelero commented:
> Would you be able to add unit tests for QSV and VAAPI in `media.service.spec.ts` to cover this case? Specifically, there needs to be a filter to convert to nv12 when the scale filter is absent.

Of course, I'm no expert but I'll do my best 🙂

andresichelero commented:
> @andresichelero would you be able to still add the unit tests?

@danieldietzler I encountered some problems trying to run the tests locally to review my implementation, and lately I haven’t had time to dedicate to this, so if someone has some free time to do this I would greatly appreciate it.
